### ISSUE TITLE
Respect mutes all for push notifications

### ISSUE THREAD

**Issue by huumn** [stackernews/stacker.news#717]
We currently *only* respect mutes for push notifications wrt replies.

**Comment by ekzyis** [comment]
> We don't respect _any_ of the notification type settings (as set in `/settings`) for push notifications afaict.

We have [this function](https://github.com/stackernews/stacker.news/blob/9698679d381e4706348c0c2263bb36090f03a01b/api/webPush/index.js#L34-L49) in _api/webPush.js_:

```js
const createUserFilter = (tag) => {
  // filter users by notification settings
  const tagMap = {
    REPLY: 'noteAllDescendants',
    MENTION: 'noteMentions',
    TIP: 'noteItemSats',
    FORWARDEDTIP: 'noteForwardedSats',
    REFERRAL: 'noteInvites',
    INVITE: 'noteInvites',
    EARN: 'noteEarning',
    DEPOSIT: 'noteDeposits',
    STREAK: 'noteCowboyHat'
  }
  const key = tagMap[tag.split('-')[0]]
  return key ? { user: { [key]: true } } : undefined
}
```

which is [called inside of `sendUserNotification`](https://github.com/stackernews/stacker.news/blob/9698679d381e4706348c0c2263bb36090f03a01b/api/webPush/index.js#L93) to filter push subscriptions by user settings:

```js
const userFilter = createUserFilter(notification.tag)
const payload = createPayload(notification)
const subscriptions = await models.pushSubscription.findMany({
  where: { userId, ...userFilter }
})
```

Is this not what you mean?

**Comment by huumn** [comment]
Oh my bad, the abstraction is so lightweight I missed it! I'll update the issue

**Comment by SatsAllDay** [comment]
I think it would be helpful to enumerate the push notifications supported today, and then build a truth table as to whether they should take mutes into account, based on the code in `lib/webPush.js`:

|Push Notification Type|Mutes should apply?|Does mute currently apply?|Required Fix?|
|-|-|-|-|
|REPLY|Yes|Yes|❌|
|MENTION|Yes|No|✅|
|FOLLOW|Yes|No|✅|
|TERRITORY_POST|Yes|No|✅|
|TERRITORY_TRANSFER|No|N/A|❌|
|TIP|No|N/A|❌|
|FORWARDEDTIP|No|N/A|❌|
|REFERRAL|No|N/A|❌|
|INVITE|No|N/A|❌|
|EARN|No|N/A|❌|
|DEPOSIT|No|N/A|❌|
|WITHDRAWAL|No|N/A|❌|
|STREAK|No|N/A|❌|
|STREAK-LOST|No|N/A|❌|

So to summarize, this issue will require updating the push notification code to honor mutes in the following scenarios:

1. Don't send a push if a stacker that you've muted, mentions you in an item.
2. Don't send a push if a stacker that you've muted, and also subscribed to, creates an item type to which you're subscribed. This one is funny, are people really muting subscribed users?
3. Don't send a push if a stacker that you've muted submits a post to a territory to which you've subscribed.

Did I get all of them? Any others? Anything wrong?

**Comment by ekzyis** [comment]
Mhh, yes, I think this makes sense. The alternative would be to allow it and overwrite the previous setting but I like explicit things and this should be an edge case so it's probably not worth thinking too much about it. A simple check in the backend and throwing an error seems best.

**Comment by SatsAllDay** [comment]
Will do! Also, please review the above table to see if I missed anything in my assessment, if you get a chance :) I have changes locally that I'll probably push for review this weekend.
